rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+        """Create a new context for use with `with`.
+
+        E.g:
+            with controller(x=3, y=4):
+                # All commands will now communicate with chip (3, 4)
+        """
+        return self.get_new_context(**context_args)
+
+    def send_scp(self, x, y, *args, **kwargs):
+        """Determine the best connection to use to send an SCP packet and use
+        it to transmit.
+
+        See the arguments for
+        :py:method:`~rig.machine_controller.scp_connection.SCPConnection` for
+        details.
+        """
+        return self.connections[0].send_scp(x, y, *args, **kwargs)
+
+    def boot(self, width, height, **boot_kwargs):
+        """Boot a SpiNNaker machine of the given size.
+
+        Parameters
+        ----------
+        width : int
+            Width of the machine (0 < w < 256)
+        height : int
+            Height of the machine (0 < h < 256)
+
+        For further boot arguments see
+        :py:func:`~rig.machine_controller.boot.boot`.  **Regardless** of the
+        values passed to this method the struct and boot data from `__init__`
+        will be used.
+
+        Notes
+        -----
+        The constants `rig.boot.spinX_boot_options` can be used to specify boot
+        parameters, for example:
+
+            controller.boot(**spin5_boot_options)
+
+        Will boot the Spin5 board connected to by `controller`.
+        """
+        boot_kwargs.update({
+            "boot_data": self.boot_data,
+            "struct_data": self.struct_data,
+        })
+        boot.boot(self.initial_host, width=width, height=height, **boot_kwargs)
+
+    @ContextMixin.use_contextual_arguments
+    def software_version(self, x=Required, y=Required, processor=0):
+        """Get the software version for a given SpiNNaker core.
+        """
+        sver = self.send_scp(x, y, processor, SCPCommands.sver)
+
+        # Format the result
+        # arg1 => p2p address, physical cpu, virtual cpu
+        p2p = sver.arg1 >> 16
+        pcpu = (sver.arg1 >> 8) & 0xff
+        vcpu = sver.arg1 & 0xff
+
+        # arg2 => version number and buffer size
+        version = (sver.arg2 >> 16) / 100.
+        buffer_size = (sver.arg2 & 0xffff)
+
+        return CoreInfo(p2p, pcpu, vcpu, version, buffer_size, sver.arg3,
+                        sver.data)
+
+    @ContextMixin.use_contextual_arguments
+    def write(self, address, data, x=Required, y=Required, p=Required):
+        """Write a bytestring to an address in memory.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start writing the data.
+        data : :py:class:`bytes`
+            Data to write into memory.
+        """
+        # While there is still data perform a write: get the block to write
+        # this time around, determine the data type, perform the write and
+        # increment the address
+        while len(data) > 0:
+            block, data = data[:256], data[256:]
```

Comment:
In principle these can vary and should be determined by call `get_software_version`... yay!